doc/bash-completion/CMakeLists.txt

```diff
@@ -0,0 +1,50 @@
+# Detects the best place to install bash completion support
+# - Windows is not supported
+# - TODO: macOS support can be provided through Homebrew
+IF(NOT LMMS_BUILD_WIN32 AND NOT LMMS_BUILD_APPLE)
+
+	# standard install location for most distributions
+	SET(BASHCOMP_FALLBACK "/usr/share/bash-completion/completions")
+
+	# check for common portable installation (CMAKE_INSTALL_PREFIX=../target)
+	GET_FILENAME_COMPONENT(BINARY_ADJ_INSTALL "${CMAKE_BINARY_DIR}/../target" ABSOLUTE)
+	IF("${BINARY_ADJ_INSTALL}" STREQUAL "${CMAKE_INSTALL_PREFIX}")
+		SET(SKIP_PKG_CHECK TRUE)
+	ENDIF()
+
+	# check for other signs of user-level install
+	SET(USER_LEVEL_INSTALL "/home/;/root/;/Users/")
+	FOREACH(_startswith ${USER_LEVEL_INSTALL})
+		IF("${CMAKE_INSTALL_PREFIX}" MATCHES "^${_startswith}")
+			SET(SKIP_PKG_CHECK TRUE)
+			BREAK()
+		ENDIF()
+	ENDFOREACH()
+
+	# determine BASHCOMP_PKG_PATH variable
+	IF(SKIP_PKG_CHECK)
+		SET(BASHCOMP_PKG_PATH "${CMAKE_INSTALL_PREFIX}${BASHCOMP_FALLBACK}")
```

Comment:
I believe it may give a wrong path because of `/usr`. For example, `~/.local` prefix should result in `~/.local/share/...`, not `~/.local/usr/share/...`.

Reply:
> ~/.local prefix should result in ~/.local/share/

Says who?  What's the difference between installing between these two unrecognized locations?  They're both rubbish to any type of bash completion.  If you want to blindly prefix, then we go back to the "can I write here?" logic and always fallback to a consistent path if we can't.

Reply:
`~/.local/share/bash-completion/completions` does work by default. `/usr` is usually the install prefix for distributon-provided packages.

Reply:
For reference: https://github.com/scop/bash-completion/blob/c2dd6745661d886327bb79591a6c6911f6f5c3f6/bash_completion#L2029-L2033

Reply:
@PhysSong it is not our place to take over a user's home directory or default settings via CMake.  CMake `INSTALL` is simply not intended for user-space installs.  It's designed for root-level POSIX installs.  If you think it's common practice to install LMMS to `~/.local/bin/lmms` please provide an example of where people would leverage CMake for deployment.  The validity of this location is irrelevant.

Reply:
>  If you think it's common practice to install LMMS to `~/.local/bin/lmms`

I saw someone does it. `pip` also installs some scripts to the same directory: https://www.python.org/dev/peps/pep-0370/
I think it may be useful for users who don't have a root access.

Reply:
We can add `~/.local` to the whitelist, it's OK for me.

Reply:
@PhysSong `pip` is a user-space application by-design.  `pip` is a package management tool.  CMake is not, it's a build tool and the `INSTALL_PREFIX` is not intended to modify the user's environment.  Please don't do this.

Reply:
> We can add `~/.local` to the whitelist, it's OK for me.

Why the hell would anyone install LMMS to `~/.local`?

Reply:
> Why the hell would anyone install LMMS to `~/.local`?

I guess adding this to the whitelist is faster than discussing it :stuck_out_tongue: Let's bring this stupid feature request to an end...

Reply:
> I guess adding this to the whitelist is faster than discussing it 😛 Let's bring this stupid feature request to an end...

Funny, but untrue.  It's not faster, it's not practical and it's not wise.  Why would anyone install an application to a hidden directory like `.local`?  The answer: They wouldn't.  Ever.

@PhysSong's talking about a special direction, not unlike the one we're hardcoding for root-level installs.  CMake needs to stay out of the user's `.local`. That's the job for [`appimaged`](https://github.com/AppImage/appimaged), and it's successor [`AppImageLauncher`](https://github.com/TheAssassin/AppImageLauncher).  This is NOT NOT NOT our job, ever with CMake unless CMake starts acting like a user-space deployment helper.  This prefix has no basis with build logic.  It just looks technically plausible.  The only argument that could be made is the manual override, which we can handle without weird whitelisting. :)

Reply:
Well, I don't mind how you decide here. I'm fine with either decision.

Reply:
Back to the leading `/usr`, it shouldn't be there when concatenating paths. I still think `${CMAKE_INSTALL_PREFIX}/share/...` is natural because it becomes `/usr/share/...` and `/usr/local/share/...` for `/usr` and `/usr/local` respectively.

Reply:
> Back to the leading /usr, it shouldn't be there when concatenating paths.

This isn't about concatenating paths.  PLease review the latest proposal.  `${CMAKE_INSTALL_PREFIX}` should never be used unless it's a non-root or non-sudo install.  `/usr` is mandatory for the [AppImages to eventually use it](https://github.com/TheAssassin/AppImageLauncher/issues/87#issuecomment-423610462).  As far as I can tell `/usr/share/bash-completion/completions` is a made-up, magic path. Let's keep it as such.  Let's not overthink the prefix because it's non-standard and it's a bad idea.

If you prefer to append `${BASHCOMP_PKG_PATH}` to the end of `${CMAKE_INSTALL_PREFIX}`, I'll take that, but this latest code proposal honors the `pkg-config` value first, hard-coded `/usr/share/bash-completion/completions` second, and if neither of those are writable, it tucks it into whatever strange prefix was used.

Prefix prediction is fine if everything follows that rule, but no evidence has been provided to suggest bash-completion follows these arbitrary prefix rules, please don't propose them, they make this PR much more difficult than it needs to be and it also breaks the (eventual) AppImages.

If you still aren't OK with that, please fix the AppImage path to match the standard as part of this PR, but post-install during the packaging stage.

Reply:
I'll clarify. We don't install the LMMS binary to `${CMAKE_INSTALL_PREFIX}/bin`, not `${CMAKE_INSTALL_PREFIX}/usr/bin`. I wanted to say that this pass calculation doesn't make sense because the `${BASHCOMP_FALLBACK}` has a leading `/usr` which is needed for absolute paths, not for relative paths.

Reply:
I double checked the appimage and the `/usr` is prefixed from the script so that should be fine as `${CMAKE_INSTALL_PREFIX}/share/`.

If that's it, @johaneslorenz can we use the new proposal?

Reply:
@tresf Without following the conversation, the code below seems to cover most cases I can think of. I can't see any errors, but if you'd make a commit I could run some tests.

Btw, `ugo+r` should be identical to `+r` ?

Reply:
Yes, `ugo` means everyone.

Reply:
> Btw, `ugo+r` should be identical to `+r`?

No, `+r` is identical to `u+r`. `ugo+r` is identical to the cmake's `PERMISSIONS OWNER_READ GROUP_READ WORLD_READ`.